src/julia-syntax.scm

```diff
@@ -693,46 +693,52 @@
                 ,@locs
                 (call (curly ,name ,@params) ,@field-names)))))
 
-(define (new-call Tname type-params params args field-names field-types)
+(define (new-call Tname type-params sparams params args field-names field-types)
   (if (any kwarg? args)
       (error "\"new\" does not accept keyword arguments"))
   (if (length> params (length type-params))
       (error "too few type parameters specified in \"new{...}\""))
-  (let ((Texpr (if (null? type-params)
-                   `(outerref ,Tname)
-                   `(curly (outerref ,Tname)
-                           ,@type-params))))
+  (if (length> type-params (length params))
+      (error "too many type parameters specified in \"new{...}\""))
+  (let* ((Texpr (if (null? type-params)
+                    `(outerref ,Tname)
+                    `(curly (outerref ,Tname)
+                            ,@type-params)))
+         (tn (make-ssavalue))
+         (field-convert (lambda (fld fty val)
+                          (if (equal? fty '(core Any))
+                              val
+                              `(call (top convert)
+                                     ,(if (and (equal? type-params params) (memq fty params) (memq fty sparams))
```

Comment:
I don't think the `(memq fty sparams)` check is necessary.

Reply:
I actually added a test for that. This is stricter than necessary (we only really need to show that duplicating `fty` won't have side-effects here—or we needed to introduce an ssavalue), but it seems like not a very common case anyways.